Ovn k kind updates

astoycos opened the pull request:
This PR works to close #703 

It fixes the makefile target `make clusters using=ovn` and then ensures we use the work upstream in ovn-k. This cannot be merged until https://github.com/ovn-org/ovn-kubernetes/pull/2747 goes in 

Depends on https://github.com/ovn-org/ovn-kubernetes/pull/2747
 
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

astoycos committed:
Fix `make clusters using=ovn`

a1ab011f2271478a694aeba864099f09f37ff93d broke
being able to specify a cable driver because
the setting were populated before the `using` var
was parsed from the command line

This commit ensures we choose the correct settings
yaml based on what is specified in the `using` flag.
By parsing it early in `Makefile.dapper` and passing
it to the dapper build env in order to ensure it is
always up to date.

Signed-off-by: Andrew Stoycos <astoycos@redhat.com>

astoycos commented:
Still waiting on https://github.com/ovn-org/ovn-kubernetes/pull/2747 to merge :(

astoycos commented:
Note: I still need to update the git submodule to point to upstream ovn-k rather than the submariner fork here.

skitt commented:
> Note: I still need to update the git submodule to point to upstream ovn-k rather than the submariner fork here.

If we can use upstream ovn-k directly now, do we still need a submodule?

astoycos commented:
Well we still need to source the ovn-k kind script @skitt, what's the "most official" way to do that? It's either a submodule or download [the script](https://github.com/ovn-org/ovn-kubernetes/blob/master/contrib/kind.sh) manually every time we need it right? (unfortunately it's not packaged in the ovn-k image)

skitt commented:
> Well we still need to source the ovn-k kind script @skitt, what's the "most official" way to do that? It's either a submodule or download [the script](https://github.com/ovn-org/ovn-kubernetes/blob/master/contrib/kind.sh) manually every time we need it right? (unfortunately it's not packaged in the ovn-k image)

I vote for downloading it when we need it; that seems preferable to cloning a 103MiB repository even when we don’t need OVN.

astoycos commented:
++ I agree will implement this here,  thanks!

skitt commented on Makefile.inc:
```diff
@@ -1,5 +1,4 @@
 , := ,
-_using = $(subst $(,), ,$(using))
```
`Makefile.inc` is shared across all projects, this will break any project relying on `_using` (`submariner`, `submariner-charts`, `submariner-operator`).

astoycos replied:
Ok so I guess what I probably need to do is just ensure that in the MAKEFILE we still `include MAKEFILE.inc` so that we still parse the `using=` flag before firing up the clusters, let me make sure this works

astoycos replied:
Fixed

skitt commented on package/Dockerfile.shipyard-dapper-base:
```diff
@@ -5,10 +5,10 @@ FROM fedora:35
 ENV DAPPER_HOST_ARCH=amd64 SHIPYARD_DIR=/opt/shipyard SHELL=/bin/bash \
     DAPPER_RUN_ARGS="--net=kind"
-ENV HOST_ARCH=${DAPPER_HOST_ARCH} ARCH=${DAPPER_HOST_ARCH} PATH=/go/bin:/usr/local/go/bin:$PATH \
+ENV HOST_ARCH=${DAPPER_HOST_ARCH} ARCH=${DAPPER_HOST_ARCH} PATH=/root/.local/bin:/go/bin:/usr/local/go/bin:$PATH \
```
Do we need this? `kind.sh` is executed without relying on `PATH`; is anything else added to `/root/.local/bin`?

astoycos replied:
> kind.sh is executed without relying on PATH 

Is it? When I remove that dir from PATH the kind script cannot use the `j2` binary which is installed by DNF at `/root/.local/bin/j2`  in the dapper build container.

skitt replied:
`kind.sh` is run as `./kind.sh` so there’s no `PATH` lookup for that; but `j2` is the answer to my question.

skitt commented on Makefile:
```diff
@@ -13,6 +13,8 @@ ifneq (,$(DAPPER_HOST_ARCH))
 
 # Running in Dapper
 
+include Makefile.inc
```
If we need this here, line 26 should be removed (but I’m not sure why it would make a difference).

astoycos replied:
Ah my bad.... And I think it makes the difference since currently we decide on the settings yaml (OVN `.shipyard.e2e.ovn.yml` / Standard `$(DAPPER_SOURCE)/.shipyard.e2e.yml`)  BEFORE we actually import the value of `using` from `Makefile.inc`....

skitt replied:
Ah yes, I missed the reference to `_using` just below. So we need this line here, and the `include` further down removed.

astoycos replied:
Fixed!

astoycos commented:
Huh @skitt That ^ breaks things because now we go and call `./clusters.sh` before populating the SETTINGS var  :( . 

We have a bit of a circular dependency going on here, why can't we just move 
```
ifneq (,$(filter ovn,$(_using)))
SETTINGS ?= $(DAPPER_SOURCE)/.shipyard.e2e.ovn.yml
else
SETTINGS ?= $(DAPPER_SOURCE)/.shipyard.e2e.yml
endif
``` 

to `makefile.inc`, since that's where we actually call the clusters script AND parse the `using` flag?

astoycos committed:
Update OVN-K image, download kind script

Fixes #703

This patch ensures that the upstreamed submariner kind hacks
for ovn-kubernetes are properly used, and ensures we pull
the ovn-k image from the correct location

it also removes the ovn-kubernetes submodule since all we really
need is the OVN-K kind scripts and templates which we now download
on the fly from ovn-kubernetes master.

Signed-off-by: Andrew Stoycos <astoycos@redhat.com>

astoycos commented:
The `[Deploy (ovn)]` Job will fail until the dependent PR merges

mkolesnik commented on Makefile.inc:
```diff
@@ -59,6 +59,9 @@ endif
 
 ifneq (,$(filter ovn,$(_using)))
 PRELOAD_IMAGES += submariner-networkplugin-syncer
+SETTINGS ?= $(DAPPER_SOURCE)/.shipyard.e2e.ovn.yml
+else
+SETTINGS ?= $(DAPPER_SOURCE)/.shipyard.e2e.yml
```
Please keep them where they were, there's no guarantee consuming projects will have exact same file names

astoycos replied:
Thanks for the look @mkolesnik! Just a quick question, If you look at my comment [here](https://github.com/submariner-io/shipyard/pull/704/commits/7ba768f0453ecf3827ae400ff2815986ae713811) you can see I tried to change this bit because  it seemed to break `make clusters using=ovn` since we don't populate the `using` flag internally until after picking which settings file to use....Do you have any recs on how to best fix this?

astoycos replied:
I fixed this by parsing the `using` flag in both makefiles so that we can ensure we're always up to date. I am definitely not a MAKE expert so please let me know if this makes sense.

mkolesnik replied:
Yes there's a bit of a "chicken and egg" situation here since the `_using` part is traditionally parsed in `Makefile.inc` but `SETTINGS` is expected to be set before including it, so we were actually silently not using `ovn` anywhere for a while.

Perhaps the best solution is to move `_using` parsing into `Makefile.dapper` as that one gets loaded way before, or just create some dedicated `Makefile.variables` file (to load any such shared variables).

astoycos replied:
Done, I added the parsing of the `using` flag to `Makefile.dapper` I think it makes it much easier to understand

astoycos commented:
Dependent OVN-K PR has merged!

astoycos commented:
So in the OVN test the cluster now deploys successfully in CI, however the connectivity tests still fail, this should be fixed by https://github.com/submariner-io/submariner/pull/1689

mkolesnik commented on Makefile:
```diff
@@ -15,6 +15,10 @@ ifneq (,$(DAPPER_HOST_ARCH))
 
 # Running in Dapper
 
+# parse `using` flag here and in makefile.inc or the incorrct kind config yaml
```
```suggestion
# parse `using` flag here and in makefile.inc or the incorrect kind config yaml
```

nit

astoycos replied:
Done

mkolesnik commented on scripts/shared/clusters.sh:
```diff
@@ -304,6 +302,16 @@ function warn_inotify() {
+function download_ovnk() {
+    if [[ "${cluster_cni[$cluster]}" == "ovn" ]]; then
+        rm -rf ovn-kubernetes/
+        # To deploy ovn-kubernetes properly it's easiest to just clone the
```
```suggestion

        # To deploy ovn-kubernetes properly it's easiest to just clone the
```
nit

astoycos replied:
Updated this comment

mkolesnik commented on scripts/shared/clusters.sh:
```diff
@@ -304,6 +302,16 @@ function warn_inotify() {
+        rm -rf ovn-kubernetes/
+        # To deploy ovn-kubernetes properly it's easiest to just clone the
+        # ovn-kubernetes repo from master on the fly in order to get the required
+        # kind scripts, and manfest generation templates.
```
```suggestion
        # kind scripts, and manifest generation templates.
```
nit

astoycos replied:
Done

mkolesnik commented on scripts/shared/clusters.sh:
```diff
@@ -304,6 +302,16 @@ function warn_inotify() {
 }
 
+function download_ovnk() {
+    if [[ "${cluster_cni[$cluster]}" == "ovn" ]]; then
```
This won't work properly since you're not calling it in a context where `$cluster` is set.

You should either call it from `deploy_kind_ovn` and risk parallel process forks overwriting each other, or check if any of the clusters needs ovn (basically if `${cluster_cni[*]}` contains `ovn`).

Also fail fast is easier to read so if no cluster needs ovn you can `return` from the function, otherwise do what you need.

astoycos replied:
right so calling it in deploy_kind_ovn creates issues, I went ahead and checked if ${cluster_cni[*]} contains "ovn" and made it so that we fail fast if possible

mkolesnik reviewed (changes requested):
Thanks for you work on this, some nits and one big issue :)

astoycos commented:
I think this _should_ be ready to go @mkolesnik, PTAL if you get the chance :)

mkolesnik commented on scripts/shared/clusters.sh:
```diff
@@ -304,13 +302,28 @@ function warn_inotify() {
+# ovn-kubernetes repo from master in order to access the required
+# kind scripts, and manifest generation templates.
+function download_ovnk() {
+    if [[ ${cluster_cni[*]} != *"ovn"* ]]
+    then
```
```suggestion
    if [[ "${cluster_cni[*]}" != *"ovn"* ]]; then
```

Some nits :)

astoycos replied:
+1 done

mkolesnik commented on scripts/shared/clusters.sh:
```diff
@@ -304,13 +302,28 @@ function warn_inotify() {
+    fi
+
+    echo "Cloning ovn-kubernetes from source"
+    rm -rf ovn-kubernetes/
```
Should probably add `ovn-kubernetes` to `.gitignore`?

astoycos replied:
Yep :)

mkolesnik commented on scripts/shared/clusters.sh:
```diff
@@ -304,13 +302,28 @@ function warn_inotify() {
+    fi
+
+    echo "Cloning ovn-kubernetes from source"
+    rm -rf ovn-kubernetes/
+    git clone https://github.com/ovn-org/ovn-kubernetes.git
```
```suggestion
    git clone https://github.com/ovn-org/ovn-kubernetes.git \
        || { git -C ovn-kubernetes fetch && git -C ovn-kubernetes reset --hard origin/master; }
```

On local envs this might be preferrable instead of cloning each time?

We could also use `git clone https://github.com/ovn-org/ovn-kubernetes.git || git -C ovn-kubernetes pull` for simplicity but if the remote branch diverges (e.g. force push) then it would fail, though on GH it's unlikely but is a **very remote** possibility.

astoycos replied:
Yep, went with the first option, I doubt it will break anytime soon

mkolesnik reviewed (commented):
Thanks, some thoughts and some nits :)

skitt commented on scripts/shared/clusters.sh:
```diff
@@ -304,13 +302,28 @@ function warn_inotify() {
+    fi
+
+    echo "Cloning ovn-kubernetes from source"
+    rm -rf ovn-kubernetes/
```
I think part of @mkolesnik’s comment was that we shouldn’t unconditionally delete `ovn-kubernetes`, to avoid having to re-download the full git repository all the time when building locally. That’s how the suggested `git clone ... || git reset` works: if `ovn-kubernetes` already exists, `git clone` will fail, and instead `git fetch && git reset` will update the existing clone.

astoycos replied:
Ah yep of course, sorry about this, fixed!

skitt commented on scripts/shared/clusters.sh:
```diff
@@ -304,13 +302,28 @@ function warn_inotify() {
+# ovn-kubernetes repo from master in order to access the required
+# kind scripts, and manifest generation templates.
+function download_ovnk() {
+    if [[ ${cluster_cni[*]} != *"ovn"* ]]; then
```
I’m not sure this does what is needed, I’m checking.

dfarrell07 commented:
I see the dependent PR is merged. Do we expect `make deploy using=ovn` to work at this point? It seems to be consistently failing.

I see you mention this getting `make clusters using=ovn` working. We could add a test for that here if you think that would be helpful: https://github.com/submariner-io/shipyard/blob/devel/.github/workflows/scripts.yml#L19

astoycos commented:
So I believe the deployment is coming up correctly @dfarrell07 However the connection test is failing 

```
Max attempts reached, failed to run 'test_connection netshoot-664cc4c98b-56nb8 100.1.48.194'!
``` 

This test will not pass until https://github.com/submariner-io/submariner/pull/1689 has merged. 

> I see you mention this getting make clusters using=ovn working. We could add a test for that here if you think that would be helpful: https://github.com/submariner-io/shipyard/blob/devel/.github/workflows/scripts.yml#L19 

100% I'll look at making a test for that, verifying that the actual CNI in use is OVN-K not weave

dfarrell07 commented:
> This test will not pass until https://github.com/submariner-io/submariner/pull/1689 has merged.

Okay thanks for clarifying. I marked this PR as dependent on that one for clarity, hope that's right.

astoycos commented:
So @dfarrell07 sorry let me try and clarify...It's a bit of a circular dependency here, we actually need this PR to land before https://github.com/submariner-io/submariner/pull/1689 because it allows the submariner+ovn-K kind setup to come up with new OVN-K images, those of which are required for https://github.com/submariner-io/submariner/pull/1689's CI to pass

dfarrell07 commented:
Okay thanks for clarifying! So it seems like we need to have both ready to merge before merging either?

astoycos commented:
> Okay thanks for clarifying! So it seems like we need to have both ready to merge before merging either?

For CI to be green everywhere yes.... however I think the only reason CI is even passing now with OVN is because it's running at an old/fixed ovn-k version so it's kind of like a false positive since on real OCP+ovn-k clusters it won't work. 

I don't know what the protocol is in the submariner community is on this but I'd rather see this get merged first to ensure that CI looks good on https://github.com/submariner-io/submariner/pull/1689.  Maybe we make the OVN job not required until 1689 lands to allow this to go in first, or something along those lines?

astoycos commented:
Further proof of the above issue -> https://github.com/submariner-io/submariner/pull/1689#issuecomment-1090887976

dfarrell07 commented:
@astoycos Okay that makes sense, agreed. Let's get this in and let the CI correctly fail, make the OVN jobs not-required.

@tpantelis That's possible with the GitHub required job config knobs, right? Basically the params to the job count as part of the job name we can filter on? This new issue with workflow names not being able to be filtered on has me doubting. Maybe it would be good to go ahead and make that config change?

@mkolesnik Please re-review :pray:

mkolesnik commented on Makefile.dapper:
```diff
@@ -3,11 +3,13 @@
 
 PROJECT ?= $(notdir $(CURDIR))
 BASE_DAPPER := Dockerfile.dapper
+USING = $(subst $(,), ,$(using))
```
Interesting approach, though I wonder if you also pass `USING` when calling from outside dapper if this will hold or break :)

(Not an issue just an observation)

astoycos replied:
> Perhaps the best solution is to move _using parsing into Makefile.dapper as that one gets loaded way before

So I believe this is the only way to parse that flag within `Makefile.dapper` and pass it into the dapper build environment, at least from what I've seen. Unfortunately the naming `_using` makes it really hard to pass the var into the build environment  since we can only do so with environment vars

mkolesnik replied:
Yeah no that's not what I meant but OK :)

I was talking about what would happen if someone would theoretically try running `make e2e USING="bla bla"` since it would be both an env var and a make flag but it was just a random thought about the weirdness of Make ;)

astoycos replied:
Ah I gotcha, yeah I'm not sure about that, we set and export USING within the Makefile so hopefully anything user provided would be overridden!

mkolesnik commented on Makefile.inc:
```diff
@@ -1,5 +1,4 @@
 , := ,
-_using = $(subst $(,), ,$(using))
```
This still can;t be removed as it;s being used by consuming projects.

They will have to be updated to use `USING` and only then can this be removed

astoycos replied:
But then we're just parsing the flag twice again..... Can we simply "translate" the variable here? i.e 

_using = ${USING} 

To ensure we only actually parse the flag a single time?

astoycos replied:
Or we need to step way back and re-evaluate again how we populate `SETTINGS` i.e Re-verting / re-working 
https://github.com/submariner-io/shipyard/commit/a1ab011f2271478a694aeba864099f09f37ff93d

Please let me know what the correct way forward is here

dfarrell07 replied:
It makes sense to me to keep the var as a "shadow" (`_using = ${USING}`) until it can be removed.

dfarrell07 replied:
Also

> /opt/shipyard/scripts/deploy.sh --cable_driver libreswan --deploytool helm --settings /go/src/github.com/submariner-io/submariner/.shipyard.e2e.yml
2022-04-12T01:42:59.0911727Z Running with: globalnet='false', deploytool='helm', deploytool_broker_args='', deploytool_submariner_args='', settings='/go/src/github.com/submariner-io/submariner/.shipyard.e2e.yml', timeout=5m, image_tag=local, cable_driver=libreswan, service_discovery=false

https://github.com/submariner-io/shipyard/runs/5982624162?check_suite_focus=true

mkolesnik replied:
Problem is all of these "using" cases in the consuming e2e are actually parsed on the Shipyard level, e.g. `using=lighthouse` in Submariner is not tested and that would break:
https://github.com/submariner-io/submariner-operator/blob/devel/Makefile#L61-L63

dfarrell07 replied:
Yeah that makes sense, unfortunately the code path for `using` in the consuming tests isn't the same as if we were running in a consuming project directly.

astoycos replied:
> It makes sense to me to keep the var as a "shadow" (_using = ${USING}) until it can be removed. 

Done and tested manually, all seems to work!

mkolesnik reviewed (changes requested):
Still have to keep `_using` until it's not used in consuming projects (no pun intended)

dfarrell07 commented:
IMHO we should make this not-required here and the main repo, merge, and then review/merge the fix in the main repo. Need help from an Owner.

Thanks @astoycos! :partying_face:

skitt committed:
Merge branch 'devel' into ovn-k-kind-updates
